### PR TITLE
feat: add lit protocol encryption / decryption

### PR DESCRIPTION
**Pull request by rodrigopavezi**
Fixes [#146 ](https://github.com/RequestNetwork/web-components/issues/146)

## Release Notes

- **New Features**
	- Enhanced wallet connection management in the Navbar component.
	- Conditional rendering of the InvoiceDashboard based on wallet connectivity status.
	- Integration of new dependencies for improved functionality.
	- Introduction of functions to facilitate interoperability between Viem Client and ethers.js.
	- New state management for decryption functionality in the context provider.
	- New properties added to the InvoiceDashboard for managing decryption settings.
	- Added a new environment variable for the Request Node in `.env.example`.
	- Updated the README to include new environment variables related to the Lit Protocol.

- **Bug Fixes**
	- Improved error handling for wallet connection initialization.

- **Documentation**
	- Updated README to reflect new environment variables and removed outdated ones.

- **Chores**
	- Removed deprecated initialization function for the Request Network.
	- Reformatted `tsconfig.json` for improved readability.

### CODE REVIEW

File: next.config.mjs [hunk 1]
```diff
@@ -1,6 +1,10 @@
 /** @type {import('next').NextConfig} */
 const nextConfig = {
   reactStrictMode: true,
+  webpack: (config) => {
+    config.externals.push("pino-pretty");
+    return config;
+  },
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply next.config.mjs:7]: It fails without it

File: utils/ethers.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  const { account, chain, transport } = client
+  const network = {
+    chainId: chain.id,
+    name: chain.name,
+    ensAddress: chain.contracts?.ensRegistry?.address,
+  }
+  const provider = new providers.Web3Provider(transport, network)
+  const signer = provider.getSigner(account.address)
+  return signer
+}
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/ethers.ts:16]: Using wagmi code here

File: components/common/Navbar.tsx [hunk 3]
```diff
@@ -43,6 +47,16 @@ const Navbar = () => {
...
 
+  useEffect(() => {
+    if (!account.isConnected) {
+      console.log("disconnecting");
+      disconnectWalletFromCypherProvider();
+    } else {
+      console.log("connected");
+      connectWalletToCypherProvider(signer, account.address);
+    }
+  });
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply components/common/Navbar.tsx:58]: I will fix this one

File: utils/context.tsx [hunk 4]
```diff
@@ -21,17 +28,156 @@ export const Provider = ({ children }: { children: ReactNode }) => {
...
+          },
+          httpConfig: {
+            getConfirmationMaxRetry: 120,
+          },
+        }),
+      ),
+    );
+  };
+
+  const initializeRequestNetwork = (walletClient: any) => {
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx:50]: I will change to unknown

File: utils/context.tsx [hunk 5]
```diff
@@ -21,17 +28,156 @@ export const Provider = ({ children }: { children: ReactNode }) => {
...
+                                            : undefined;
+            if (!paymentsSubgraphUrl) {
+              throw new Error(
+                `Cannot get subgraph client for unknown chain: ${chain}`,
+              );
+            }
+            return getTheGraphClient(chain, paymentsSubgraphUrl);
+          },
+        },
+      });
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx:131]: Could you create an issue for this as it is chore refactoring?

File: utils/context.tsx [hunk 6]
```diff
@@ -21,17 +28,156 @@ export const Provider = ({ children }: { children: ReactNode }) => {
...
+          nodeConnectionConfig: {
+            baseURL: process.env.NEXT_PUBLIC_REQUEST_NODE || "https://gnosis.gateway.request.network/",
+          },
+          httpConfig: {
+            getConfirmationMaxRetry: 360,
+          },
+        }),
+      ),
+    );
+  };
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx]: I will change

File: utils/context.tsx [hunk 7]
```diff
@@ -21,17 +28,151 @@ export const Provider = ({ children }: { children: ReactNode }) => {
...
+                                          'https://subgraph.satsuma-prod.com/e2e4905ab7c8/request-network--434873/request-payments-sepolia/api'
+                                        : chain === 'xdai'
+                                          ? process.env
+                                              .NEXT_PUBLIC_PAYMENTS_SUBGRAPH_URL_XDAI ||
+                                            'https://api.studio.thegraph.com/query/67444/request-payments-xdai/version/latest'
+                                          : chain === 'zksyncera'
+                                            ? process.env
+                                                .NEXT_PUBLIC_PAYMENTS_SUBGRAPH_URL_ZKSYNCERA ||
+                                              'https://subgraph.satsuma-prod.com/e2e4905ab7c8/request-network--434873/request-payments-zksyncera/api'
+                                            : undefined;
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx:122]: I will leave that change for another time

File: utils/context.tsx [hunk 8]
```diff
@@ -5,12 +5,19 @@ import React, {
...
+import { LitProtocolProvider } from '@requestnetwork/lit-protocol-cypher';
+import { Web3SignatureProvider } from '@requestnetwork/web3-signature';
+import { getTheGraphClient } from '@requestnetwork/payment-detection';
+import HttpDataAccess from '@requestnetwork/request-client.js/dist/http-data-access';
+import { CypherProviderTypes } from '@requestnetwork/types';
 
 interface ContextType {
   requestNetwork: RequestNetwork | null;
+  isWalletConnectedToCypherProvider: boolean;
+  connectWalletToCypherProvider: (signer: any, walletAddress: string) => void;
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx:19]: I will change to unknown

File: utils/context.tsx [hunk 9]
```diff
@@ -21,17 +28,151 @@ export const Provider = ({ children }: { children: ReactNode }) => {
...
+
+      setRequestNetwork(requestNetwork);
+    } catch (error) {
+      console.error('Failed to initialize the Request Network:', error);
+      setRequestNetwork(null);
+    }
+  };
+
+  const connectWalletToCypherProvider = async (
+    signer: any,
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx:141]: that will break things

File: components/common/Navbar.tsx [hunk 10]
```diff
@@ -43,6 +47,14 @@ const Navbar = () => {
     },
   ];
 
+  useEffect(() => {
+    if (!account.isConnected) {
+      disconnectWalletFromCipherProvider();
+    } else {
+      connectWalletToCipherProvider(signer, account.address as string);
+    }
+  });
```
[the remark this answers is not in the document]
Comment (aimensahnoun) [reply components/common/Navbar.tsx:56]: I agree with this

**Review by aimensahnoun — CHANGES_REQUESTED** [review]

File: utils/context.tsx [hunk 11]
```diff
@@ -1,37 +1,230 @@
...
+                                                .NEXT_PUBLIC_PAYMENTS_SUBGRAPH_URL_ZKSYNCERA ||
+                                              'https://subgraph.satsuma-prod.com/e2e4905ab7c8/request-network--434873/request-payments-zksyncera/api'
+                                            : undefined;
+            if (!paymentsSubgraphUrl) {
+              throw new Error(
+                `Cannot get subgraph client for unknown chain: ${chain}`,
+              );
+            }
+            return getTheGraphClient(chain, paymentsSubgraphUrl);
+          },
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx:156]: I will not change that now. Thanks

File: utils/context.tsx [hunk 12]
```diff
@@ -21,17 +28,156 @@ export const Provider = ({ children }: { children: ReactNode }) => {
...
+      initializeCypherProvider();
     }
   }, [walletClient, chainId, address, isConnected]);
 
+  useEffect(() => {
+    if (cypherProvider && isWalletConnectedToCypherProvider) {
+      initializeRequestNetwork(walletClient);
+    }
+  }, [cypherProvider, isWalletConnectedToCypherProvider]);
+
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply utils/context.tsx]: Resovled as Outdated :hourglass_flowing_sand:

File: utils/context.tsx [hunk 13]
```diff
@@ -21,17 +28,156 @@ export const Provider = ({ children }: { children: ReactNode }) => {
...
+  const connectWalletToCypherProvider = async (
+    signer: any,
+    walletAddress: string,
+  ) => {
+    if (cypherProvider) {
+      await cypherProvider.getSessionSignatures(signer, walletAddress);
+      console.log('Connected to Cypher Provider');
+      setIsWalletConnectedToCypherProvider(true);
+    }
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply utils/context.tsx]: Resolved as Outdated :hourglass_flowing_sand:

File: .env.example [hunk 14]
```diff
@@ -33,3 +33,7 @@ NEXT_PUBLIC_WALLET_CONNECT_PROJECT_ID="your-project-id"
 # NEXT_PUBLIC_PAYMENTS_SUBGRAPH_URL_ZKSYNCERA=https://subgraph.satsuma-prod.com/e2e4905ab7c8/request-network--434873/request-payments-zksyncera/api
 
 # NEXT_PUBLIC_GTM_ID=GTM-XXXXXXXX
+
+# NEXT_PUBLIC_REQUEST_NODE="request-node.com"
+#NEXT_PUBLIC_LIT_PROTOCOL_CHAIN=ethereum
+#NEXT_PUBLIC_LIT_PROTOCOL_NETWORK=datil-dev
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply .env.example]: @MantisClone lit supports other chains besides Ethereum EVM.

File: components/common/Navbar.tsx [hunk 15]
```diff
@@ -39,6 +44,28 @@ const Navbar = () => {
...
+    const handleConnection = async () => {
+      try {
+        setIsConnecting(true);
+        if (!account.isConnected) {
+          disconnectWalletFromCipherProvider();
+        }
+      } catch (error) {
+        if (error instanceof Error) {
+          console.error('Wallet disconnection error:', error.message);
+          // Add user notification here
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply components/common/Navbar.tsx:59]: Fixed it

File: utils/context.tsx [hunk 16]
```diff
@@ -1,37 +1,230 @@
+/** @format */
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx]: I thing it was from my Vscode fomatting. Removing

File: utils/context.tsx [hunk 17]
```diff
@@ -1,37 +1,230 @@
...
+                                                .NEXT_PUBLIC_PAYMENTS_SUBGRAPH_URL_ZKSYNCERA ||
+                                              'https://subgraph.satsuma-prod.com/e2e4905ab7c8/request-network--434873/request-payments-zksyncera/api'
+                                            : undefined;
+            if (!paymentsSubgraphUrl) {
+              throw new Error(
+                `Cannot get subgraph client for unknown chain: ${chain}`,
+              );
+            }
+            return getTheGraphClient(chain, paymentsSubgraphUrl);
+          },
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx:154]: I will do that now

File: utils/context.tsx [hunk 18]
```diff
@@ -4,34 +4,234 @@ import React, {
...
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx:156]: I am not refactor that now

File: utils/context.tsx [hunk 19]
```diff
@@ -4,34 +4,233 @@ import React, {
...
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply utils/context.tsx:155]: I will not refactor this now

File: README.md [hunk 20]
```diff
@@ -81,7 +81,8 @@ Before deploying, ensure you have created a `.env` or `.env.local` file in the r
 | NEXT_PUBLIC_PAYMENTS_SUBGRAPH_URL_SEPOLIA      | Subgraph URL for Sepolia             | ❌       |
 | NEXT_PUBLIC_PAYMENTS_SUBGRAPH_URL_XDAI         | Subgraph URL for Gnosis              | ❌       |
 | NEXT_PUBLIC_PAYMENTS_SUBGRAPH_URL_ZKSYNCERA    | Subgraph URL for ZkSyncEra           | ❌       |
-| NEXT_PUBLIC_GTM_ID                             | Google Tag Manager ID                | ❌       |
+| NEXT_PUBLIC_LIT_PROTOCOL_CHAIN                 | Lit Protocol Chain                   | ❌       |
+| NEXT_PUBLIC_LIT_PROTOCOL_NETWORK               | Lit Protocol Network                 | ❌       |
```
Comment (MantisClone) [inline README.md:85]: 1. NEXT_PUBLIC_GTM_ID is still a valid env var.
2. Please add NEXT_PUBLIC_REQUEST_NODE

File: tsconfig.json [hunk 21]
```diff
@@ -13,10 +17,21 @@
...
-      "@/*": ["./*"],
+      "@/*": [
+        "./*"
+      ]
     },
-    "typeRoots": ["./types.d.ts"],
+    "typeRoots": [
+      "./types.d.ts"
+    ],
+    "target": "ES2017"
```
Comment (MantisClone) [inline tsconfig.json:27]: Adding "target": "ES2017" is the only change in this file. Everything else in this file is a formatting change.

**Review by MantisClone — APPROVED** [review]

**Pull request merged**